File: backend/src/main/java/com/eagleauctioner/service/BidService.java
package com.eagleauctioner.service;

import com.eagleauctioner.entity.Bid;
import com.eagleauctioner.entity.BidHistory;
import com.eagleauctioner.dto.BidDTOs.*;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

/**
 * Primary interface for the bidding engine, coordinating validation, 
 * execution, and post-bid activities.
 */
public interface BidService {
    
    /**
     * Entry point for placing a bid on an auction lot.
     */
    Bid placeBid(UUID lotId, UUID bidderId, Long bidAmount, String ipAddress, String userAgent);
    
    /**
     * Retrieves the current leading bid for a lot.
     */
    Optional<Bid> findWinningBid(UUID lotId);
    
    /**
     * Retrieves the highest bid, considering privacy rules for sealed auctions.
     */
    Optional<Bid> findHighestBid(UUID lotId);
    
    /**
     * Returns the full bid history for a specific lot.
     */
    List<BidHistory> getBidHistory(UUID lotId);
    
    /**
     * Calculates the rank and status of a specific bidder on a lot.
     */
    RankStatusResponse getMyRank(UUID lotId, UUID bidderId);
    
    /**
     * Admin/Seller action to open and resolve winners for a sealed bid lot.
     */
    SealedBidOpeningResponse openSealedBids(UUID lotId, String actor);
}

File: src/api/bidService.ts
import { apiClient } from "./client";
import { ApiResponse } from "../types/auction";
import { 
  PlaceBidRequest, 
  PlaceSealedBidRequest,
  BidResponse, 
  BidHistoryResponse,
  RankStatusResponse,
  SealedBidOpeningResponse
} from "../types/bid";
import { API_ENDPOINTS } from "../constants";

export const bidService = {
  placeBid: async (lotId: string, data: PlaceBidRequest): Promise<BidResponse> => {
    const response = await apiClient.post<ApiResponse<BidResponse>>(API_ENDPOINTS.BID.PLACE(lotId), data);
    return response.data.data;
  },

  placeSealedBid: async (lotId: string, data: PlaceSealedBidRequest): Promise<BidResponse> => {
    const response = await apiClient.post<ApiResponse<BidResponse>>(API_ENDPOINTS.BID.PLACE_SEALED(lotId), data);
    return response.data.data;
  },

  getHighestBid: async (lotId: string): Promise<BidResponse | null> => {
    const response = await apiClient.get<ApiResponse<BidResponse | null>>(API_ENDPOINTS.BID.HIGHEST(lotId));
    return response.data.data;
  },

  getBidHistory: async (lotId: string): Promise<BidHistoryResponse[]> => {
    const response = await apiClient.get<ApiResponse<BidHistoryResponse[]>>(API_ENDPOINTS.BID.HISTORY(lotId));
    return response.data.data;
  },

  getMyRank: async (lotId: string): Promise<RankStatusResponse> => {
    const response = await apiClient.get<ApiResponse<RankStatusResponse>>(API_ENDPOINTS.BID.MY_RANK(lotId));
    return response.data.data;
  },

  openSealedBids: async (lotId: string): Promise<SealedBidOpeningResponse> => {
    const response = await apiClient.post<ApiResponse<SealedBidOpeningResponse>>(API_ENDPOINTS.BID.OPEN_SEALED(lotId));
    return response.data.data;
  }
};
